[Aung Ko Khant] iP

Ko-Khan opened the pull request:
# **Introducing Bit!**
>“Your mind is for having ideas, not holding them.” [(source)](https://dansilvestre.com/productivity-quotes)

### Bit helps you keep track of all your tasks! It's,
- text-based
- easy to learn
- ~~Fast~~ SUPER FAST to use 
 
All you need to do is,
1. Download it from [here](https://github.com/Ko-Khan/ip/releases/tag/A-Jar).
2. double-click it
3. add your tasks
4. let it manage your tasks for you :slightly_smiling_face:

And its **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] reminders (coming soon) 

If you are a java programmer, you can use it to practice java too! Here is the main function:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Ko-Khan committed:
Renamed chatbot to Bit and added skeleton code that says hi and exits..

Ko-Khan committed:
Bit now echoes commands and exits when command "bye" is entered

Ko-Khan committed:
Bit nows supports adding and listing! Type anything to add it to the list and cmd "list" to list out everything in the list!

Ko-Khan committed:
Created a new class Task. Bit can now mark and unmark tasks as complete or incomplete respectively.

Ko-Khan committed:
Added three classes todo, event and deadline that inherit from task class.

Ko-Khan committed:
Changed private of variables in task to protected instead

Ko-Khan committed:
Added cmnd todo that allows one to add that todos to list.

Ko-Khan committed:
Bit can now accept the event class in its list!

Ko-Khan committed:
Bit can now use deadlines!

Ko-Khan committed:
Implemented basic i/o testing along with some basic testcases, will add more as I progress

Ko-Khan committed:
Added minimal error detection

Ko-Khan committed:
Improved error handling for events and todos

Ko-Khan committed:
Error handing for deadlines. Error handling for basic errors complete! :)

Ko-Khan committed:
Changed list from an array to an arraylist<Task> with no regressions

Ko-Khan committed:
Fixed an issue when trying to mark an index not in use causes a crash

Ko-Khan committed:
Bit can now delete unwanted items from the list!

Ko-Khan committed:
Updated i/o testing for delete

Ko-Khan committed:
Removed exception string from error messages

Ko-Khan committed:
Add new branch as iP specify. Need new location to store data from chatbot. Add data/bit.txt

Ko-Khan committed:
Need to save tasks to hard disk.

Create a new method saveToFile(). It initialise and use a filewriter to writer to file ./data/bit.txt/

Ko-Khan committed:
Chatbot need to save tasks to harddisk.

Create file ./data/bit.txt. Can Create file if it does not exist. Store tasks in bit.txt.

Ko-Khan committed:
Need to save mark/unmarked status in harddisk.

Code now remembers status by saving it to bit.txt

Ko-Khan committed:
Dates entered alongside deadline need to be treat as dates and not strings.

Import LocalDate library and use it to parse dates in correct format yyyy-mm-dd and print them in dd MMMM yyyy. Add error detection and notification for incorrect date formats.

Ko-Khan committed:
Automated i/o testing is outdated after changes to deadline class.

Update input.txt and Expected.txt accordingly.

Ko-Khan committed:
Want to increase use of OOP paradigm in project.

Solution: Add classes ui, parser, storage and tasklist. Class ui  handles sending messages to user

Ko-Khan committed:
Problem: Need to use OOP more.
Solution: Add class Tasklist that handles all operations on collection of tasks.

Ko-Khan committed:
Need to make new class handling storage of tasks in harddisk for oop.
Add Storage which handles all methods and variables regarding saving to harddisk

Ko-Khan committed:
Finish class Parser which parses through inputs and return strings containing the respective commands.

Ko-Khan committed:
Revert "Finish class Parser which parses through inputs and return strings containing the respective commands."

This reverts commit 16731f4e8a543ea4f66bc6906fee211d0d187fb4.

Ko-Khan committed:
Revert "Revert "Finish class Parser which parses through inputs and return strings containing the respective commands.""

This reverts commit 850b41024304cb76a7974a5f5f392f02434ce7ec.

Ko-Khan committed:
Create package bit containing all classes related to duke board, as minimal expectations.

Ko-Khan committed:
Merge branch 'add-gradle-support'

Ko-Khan committed:
Need to implement Junit testing.
Create 2 classes TasklistTest and ParserTest. TaskListtest has 1 method testAddTo() that tests addTo() and ParserTest has 1 method testParse() that tests parse()

Ko-Khan committed:
Adjust code in build.gradle for bit specifically

Ko-Khan committed:
Add javadocs documentation for various public methods

Ko-Khan committed:
Fix some coding style errors.

Ko-Khan committed:
Implement find function.

Ko-Khan committed:
Merge branch 'branch-A-CodingStandard' into branch-Level-9

Ko-Khan committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9

Ko-Khan committed:
Update some styling errors I miss. Add documentation for new methods relating to find command.

Colex2000 commented on src/main/java/bit/Bit.java:
```diff
@@ -0,0 +1,62 @@
+import  java.util.Scanner;
+
+
+
```
Perhaps you can add header comment here 😁 also, maybe too many spaces here?

Colex2000 commented on src/main/java/bit/Bit.java:
```diff
@@ -0,0 +1,62 @@
+        while(true) {
+            UI.printLineBreak();
+            String input = scanner.nextLine();
+            String s = parser.parse(input);
```
Perhaps instead of 's', can consider other names for easy understanding 👍

Colex2000 commented on src/main/java/bit/Bit.java:
```diff
@@ -0,0 +1,62 @@
+                    System.out.println(e.getMessage());
+                }
+            }
+
```
Can consider deleting these extra spaces for consistency 😆

Colex2000 replied:
Noticed it in other areas as well

Colex2000 commented on src/main/java/bit/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+import java.time.DateTimeException;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+public class Deadline extends Task {
```
Similarly, perhaps can include comment header for classes and public methods :)

Ko-Khan replied:
I will try and add them ASAP

Colex2000 commented on src/main/java/bit/Parser.java:
```diff
@@ -0,0 +1,89 @@
+    private String word = "";
+    private Ui UI = new Ui();
+
+    public Parser() {
```
I think can exclude this as there is no body implementation 👍

Colex2000 commented on src/main/java/bit/Storage.java:
```diff
@@ -0,0 +1,170 @@
+package bit;
+
+import java.io.*;
```
Try not to import all haha. Only import the ones you need 😀

Ko-Khan replied:
Thanks, I will get rid of the wildcard soon

Colex2000 commented on src/main/java/bit/Storage.java:
```diff
@@ -0,0 +1,170 @@
+     * Save all tasks in the list to a text file.
+     * @param tasklist the tasks to be stored.
+     */
+    public   void saveAll(Tasklist tasklist) {
```
I think you accidentally added extra spaces after 'public'. Also saw it in other methods too :)

Ko-Khan replied:
I don't know how that happened lol. 😅

Colex2000 commented on src/test/java/bit/ParserTest.java:
```diff
@@ -0,0 +1,22 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class ParserTest {
+    @Test
+    public void  testParse() {
```
Good job testing most / all of the Parser class methods!

Colex2000 reviewed (commented):
Overall, pretty good work there! I think some things to note: you have some inconsistency in terms of spacing! It could be caused by accident or IDE issues haha.

Ko-Khan committed:
Add checkstyle for java

Ko-Khan committed:
Fix check style errors

There are many methods without java docs or errors in style.

Fixing them improves code quality

Ko-Khan committed:
Add new classes Main Window, Dialog Box and Launcher

Adding java FX to improve chatbot.

Ko-Khan committed:
Add ability to launch GUI from main class

Upgrade bit from terminal to a GUI. Add Main and launcher to enable GUI.
Bit class is still a stim.

Ko-Khan committed:
Add ability for chatbot to manage tasks

Chatbot's responses is a stim.

Solution: Restructure previous code so that they can be send as messages in GUI.

Ko-Khan committed:
Fix storage for Bit

Move to GUI causes problem with previous code that I use to store tasks.

Solution: Bit now saves tasks to file whenever a change is made to task list.

Ko-Khan committed:
Make some adjustments to dialogue box settings

Dialogue box is a little ugly. So change a bit to make less ugly.

Ko-Khan committed:
Make change to build gradle file

Need to make jar file. So make changes to gradle file to facilitate that.

Ko-Khan committed:
Need to test assumptions in code

Activate assertions and add some to test some basic assertions.

Ko-Khan committed:
Parse method is too long

Create helper methods parseForFind, parseForDelete, parseForMark to cut down on the length of method. This also helps avoid arrowhead structure.

Ko-Khan committed:
Method loadFile is too long

Split some chunks off into helper methods to cut down on the length.

Ko-Khan committed:
Add some assertions for areas of code in loadFile that should not be reached

Ko-Khan committed:
Merge pull request #2 from Ko-Khan/branch-A-Assertions

Branch a assertions

Ko-Khan committed:
Merge branch 'master' into branch-A-CodeQuality

Ko-Khan committed:
Merge pull request #3 from Ko-Khan/branch-A-CodeQuality

Branch a code quality

Ko-Khan committed:
Add form of reminders

User can now enter command remind [number of days tasks due in/int]. This will return all deadlines due in that many days!

Ko-Khan committed:
Using remind with an non integer causes program to crash.

Use a try catch block to catch situation where a non integer is used. Bit reminds user to enter a number instead.
Issue with gui shortening messages if they are too long. Modify dialog box to fix this.

Ko-Khan committed:
Java GUI is cutting out last line of messages

Decrease size of font and deactivate wraparound on label

Ko-Khan committed:
If statement in listout method tests assumption that task in list is not null

This makes happy path unclear and causes more indentation. To solve this, replace if statement with assertion.

Ko-Khan committed:
Ui class has many code quality issues

This may have an adverse effect on readability. So rectify this problems.

Some methods like listHits() and listDueSoon() have too many levels of indentation. Solution is to extract some of the if statements out as helper methods.
Happy path is unclear in some methods due to if else statements checking for assumptions, so use assert instead.

Ko-Khan committed:
Localize localDate now method to Singapore time

Ko-Khan committed:
Add screenshot of gui as Ui.png

Ko-Khan committed:
Update user guide

Ko-Khan committed:
Update java documentation to follow coding standard

Ko-Khan committed:
addTo method in tasklist is too long

Extract helper methods to shorten it.

Ko-Khan committed:
The test method in TasklistTest was too long

Extract out helper methods to shorten it.

Ko-Khan committed:
Update java documentation

Ko-Khan committed:
Minor changes to javadocs

Ko-Khan committed:
Make small update to user guide

Ko-Khan committed:
Remove example template

Ko-Khan committed:
Fix minor errors in user guide
